### PR TITLE
Implement more vectorized aggregate functions

### PR DESCRIPTION
**Pull request by akuzm**
Vectorize common aggregate functions like `min`, `max`, `sum`, `avg`, `stddev`, `variance` for arithmetic types, for no grouping and grouping on segmentby columns.

Tsbench shows up to 11x improvement and 2x on average on affected queries: https://grafana.ops.savannah-dev.timescale.com/d/fasYic_4z/compare-akuzm?orgId=1&var-branch=All&var-run1=3730&var-run2=3728&var-threshold=0&var-use_historical_thresholds=true&var-threshold_expression=2.5%20%2A%20percentile_cont%280.90%29&var-exact_suite_version=false&from=now-2d&to=now

Depends on:
- [x] https://github.com/timescale/timescaledb/pull/7199
- [x] https://github.com/timescale/timescaledb/pull/7147

### CODE REVIEW

**Review by erimatnor — COMMENTED** [review]
Submitting the comments I have so far. Haven't finished a full review, but I haven't found anything major yet. But I figured I shouldn't sit on this for too long, so let's start with some feedback.

File: tsl/src/nodes/vector_agg/function/float48_accum_types.c [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+#include "float48_accum_single.c"
+
+#define PG_TYPE FLOAT8
+#define CTYPE double
+#define CTYPE_TO_DATUM Float8GetDatum
+#define DATUM_TO_CTYPE DatumGetFloat8
+#include "float48_accum_single.c"
+
+#undef AGG_NAME
+#undef NEED_SXX
```
[the remark this answers is not in the document]
Comment (akuzm) [reply tsl/src/nodes/vector_agg/function/float48_accum_types.c:26]: They are defined in `float48_accum_templates.c` before including these files. Generally, the idea is that the template file undefines the macros that it is parameterized by. The `float48_accum_types.c` is parameterized by these two macros to generate the two families of transition functions that either require or not the Sxx state. I'm following this approach for template files in vectorized filters as well. Not sure where to best describe this.

File: tsl/src/nodes/vector_agg/function/int24_avg_accum_single.c [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+	.agg_emit = int24_avg_accum_emit,
+	.agg_const = FUNCTION_NAME(const),
+	.agg_vector = FUNCTION_NAME(vector),
+};
+
+#endif
+
+#undef PG_TYPE
+#undef CTYPE
+#undef DATUM_TO_CTYPE
```
[the remark this answers is not in the document]
Comment (akuzm) [reply tsl/src/nodes/vector_agg/function/int24_avg_accum_single.c:46]: Same comment as above, there's only a single instantiation of this template for a particular type, so the type is un-defined at the end, but multiple instantiations for the particular aggregate (AVG), so that one is un-defined above.

File: tsl/src/nodes/vector_agg/function/functions.c [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+static void
+count_any_vector(void *agg_state, const ArrowArray *vector, const uint64 *filter,
+				 MemoryContext agg_extra_mctx)
+{
+	CountState *state = (CountState *) agg_state;
+	const int n = vector->length;
+	const uint64 *restrict validity = (uint64 *) vector->buffers[0];
+	/* First, process the full words. */
+	for (int i = 0; i < n / 64; i++)
```
[the remark this answers is not in the document]
Comment (akuzm) [reply tsl/src/nodes/vector_agg/function/functions.c:83]: They still have the uint64 support, so it should work just as well. Probably slower, I didn't test.

File: tsl/src/nodes/vector_agg/function/float48_accum_single.c [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+#endif
+
+static pg_attribute_always_inline void
+FUNCTION_NAME(vector_impl)(void *agg_state, int n, const CTYPE *values, const uint64 *valid1,
+						   const uint64 *valid2, MemoryContext agg_extra_mctx)
+{
+	/*
+	 * Vector registers can be up to 512 bits wide.
+	 */
+#define UNROLL_SIZE ((int) (512 / 8 / sizeof(CTYPE)))
```
[the remark this answers is not in the document]
Comment (akuzm) [reply tsl/src/nodes/vector_agg/function/float48_accum_single.c:199]: This is just a guesstimate that is specific to a particular unrolled loop, no deep meaning. So I'd keep it local. Maybe we can make it more architecture-specific, for now I just wrote a "least-effort" version that is vectorized at least somehow.

File: tsl/src/nodes/vector_agg/function/sum_float_single.c [hunk 5]
```diff
@@ -0,0 +1,100 @@
+/*
```
[the remark this answers is not in the document]
Comment (akuzm) [reply tsl/src/nodes/vector_agg/function/sum_float_single.c:1]: The naming follows the respective Postgres aggregate transition functions.

File: tsl/src/nodes/vector_agg/function/float48_accum_single.c [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+	{
+		return;
+	}
+
+	/*
+	 * This code follows float8_accum(), see the comments there.
+	 */
+#ifdef NEED_SXX
+	Assert(*N > 0.0);
+	const double tmp = newval * (*N + 1.0) - (*Sx + newval);
```
[the remark this answers is not in the document]
Comment (akuzm) [reply tsl/src/nodes/vector_agg/function/float48_accum_single.c]: Done.

File: tsl/src/nodes/vector_agg/function/float48_accum_single.c [hunk 7]
```diff
@@ -0,0 +1,307 @@
...
+)
+{
+	const CTYPE newval = values[row];
+	if (!arrow_row_both_valid(valid1, valid2, row))
+	{
+		return;
+	}
+
+	/*
+	 * This code follows float8_accum(), see the comments there.
```
[the remark this answers is not in the document]
Comment (akuzm) [reply tsl/src/nodes/vector_agg/function/float48_accum_single.c]: Added the comment.

File: tsl/src/nodes/vector_agg/function/functions.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (erimatnor) [reply tsl/src/nodes/vector_agg/function/functions.c:83]: Yes, the compiler has uint64 support, but it is not a native word-length type on those machines so they need to emulate 64-bit integers.

We do build for 32 bit, but it is probably not a platform we optimize for in terms of performance. It might be good to point out that we currently don't care about optimizing for 32 bit specifically.

File: tsl/src/nodes/vector_agg/function/float48_accum_single.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (erimatnor) [reply tsl/src/nodes/vector_agg/function/float48_accum_single.c:203]: I am just wondering if there's a standard header file we can import that tells us AVX-512 is available and there some "define" related to the unroll size that says "512".

**Review by erimatnor — APPROVED** [review]
Approving. This seems mostly like boilerplate stuff. I didn't deeply check the correctness of the aggregate function implementations. I guess tests should reveal any issue there.

The macro and pre-processor stuff is a little bit hard to follow sometimes. I am wondering if that could be cleaned up, but I guess it is what we have for now.

**Review by svenklemm — APPROVED** [review]

File: tsl/src/nodes/vector_agg/function/float48_accum_single.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (akuzm) [reply tsl/src/nodes/vector_agg/function/float48_accum_single.c:203]: Not sure if there's an easy cross-platform way to do it. Generally we want to unroll to the biggest available size across architectures, if there's only a smaller width available it will still lead to vectorizable code.
